chainable/__init__.py

```diff
@@ -0,0 +1,46 @@
+from functools import reduce, _initial_missing
+from typing import TypeVar, Callable, overload, Generic
+from collections.abc import Iterable, Iterator
+
+__all__ = ["Chainable"]
+
+T = TypeVar("T")
+S = TypeVar("S")
+
+def flat_map_helper(self: "Chainable[T]", mapper: Callable[[T], Iterable[S]]) -> Iterator[S]:
+    for val in self:
+        yield from mapper(val)
+    
+
+class Chainable(Generic[T]):
+    def __init__(self, it: Iterable[T]) -> None:
+        self.it = it
+
+    def __iter__(self) -> Iterator[T]:
+        return iter(self.it)
+
+    def map(self, mapper: Callable[[T], S]) -> "Chainable[S]":
+        return Chainable(map(mapper, self))
+    
+    def filter(self, predicate: Callable[[T], bool]) -> "Chainable[T]":
+        return Chainable(filter(predicate, self))
+    
+    @overload
+    def reduce(self, reducer: Callable[[T, T], T]) -> T:
+        ...
+    
+    @overload
+    def reduce(self, reducer: Callable[[T, T], T], initial: T) -> T:
+        ...
+
+    def reduce(self, reducer, initial = _initial_missing) -> T:
+        return reduce(reducer, self, _initial_missing)
+
+    def any(self, pred: Callable[[T], bool]) -> bool:
```

Comment:
Maybe pred should default to identity (as it does for `any(self)`)?

```suggestion
    def any(self, pred: Callable[[T], Any] = lambda x: x) -> bool:
```

Reply:
That way, someone could call `Chainable(things).any()`.

With this change, I'm not sure the type signature is correct. It might need to be something like:

```suggestion
    def any(self, pred: Callable[[T], Any]) -> bool:
```

or

```suggestion
    def any(self, pred: Callable[[T], ObjectWith__bool__]) -> bool:
```

Reply:
It appears there's no good answer for "[how to annotate SupportsBool](https://stackoverflow.com/q/57712347/70170)".

Reply:
What about:
```py
class SupportsBool(Protocol):
    def __bool__(self) -> bool: ...
```

Reply:
According to [this answer](https://stackoverflow.com/a/57747466/70170), that won't work for certain (common) objects like strings and lists and integers that rely on the fallback to `__len__`. I'd probably just rely on `Any` until the typing system has formalized support for `SupportsBool` that reflects the implementation.

Reply:
`Any` does seem good enough.

Reply:
I think it's better to use `object` instead of `Any` here. `object` and its subtypes are "supports bool" by default and `Callable` is covariant in return type.

Reply:
@hoodmane

Reply:
I don't really understand what `object` means as a type. We certainly want to accept `int`, `float`, `str`. Are they `object`s?

Reply:
> I don't really understand what `object` means as a type. We certainly want to accept `int`, `float`, `str`. Are they `object`s?

Yes, they are. All Python objects inherit from `object`.
That said, `Callable[[T], object]` will allow others to pass in a function that takes a single object of type `T` (or supertypes) and returns any object. Up to that point there is no difference. The difference here is that using `object` protects us: for instance, if we were to access an inexistent attribute of the return value of `pred`, when annotated with `Any` that would not produce any warnings, contrary to `pred` return value annotated as `object`, which ensures that we only access [the `object` interface](https://github.com/python/typeshed/blob/fd67f0a457a601ddb916dc35dfcdbda30b363f08/stdlib/builtins.pyi#L96-L128) and nothing else.